File: 141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/*
 * Approach 1: Testing NULL pointer in a limited time, say 0.5s.
 * Then we think it has a loop after 0.5s when current node still
 * not meet NULL pointer.
 */
/* Approach 2: Hashtable or set. */
/* Approach 3: Two pointers. */
bool hasCycle(struct ListNode *head) {
    struct ListNode *slow, *fast;

    slow = fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

File: Python3/141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        cur = head
        d = set()

        while cur:
            if cur in d:
                return True
            else:
                d.add(cur)
                cur = cur.next
        return False

    def hasCycle1(self, head):
        """
        :type head: ListNode
        :rtype: bool
        Two pointers solution
        """
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        return False
